File: p7/wis.c
#include <stdio.h>
#include <stdlib.h>

// a structure to represent a job
struct Job
{
    int start;
    int end;
    int weight;
};

// a function to sort the jobs based on their end time
void sortJobs(struct Job *jobs, int numJobs)
{
    // quick sort
    // base case
    if (numJobs <= 1)
    {
        return;
    }

    // choose the pivot
    int pivot = jobs[numJobs / 2].end;

    // partition
    int i = 0;
    int j = numJobs - 1;
    while (i <= j)
    {
        while (jobs[i].end < pivot)
        {
            i++;
        }
        while (jobs[j].end > pivot)
        {
            j--;
        }
        if (i <= j)
        {
            struct Job temp = jobs[i];
            jobs[i] = jobs[j];
            jobs[j] = temp;
            i++;
            j--;
        }
    }

    // recursively call on the left and right subarrays
    sortJobs(jobs, j + 1);
    sortJobs(jobs + i, numJobs - i);
}

// a function to determine the maximum weight of jobs that can be scheduled using dynamic programming
void maxWeight(int numJobs, struct Job *jobs, long long *maxWeightArray)
{
    for (int i = 1; i <= numJobs; i++)
    {
        // find the latest previous job that does not overlap with the current job
        int j = i - 1;
        while (j > 0 && jobs[j - 1].end > jobs[i - 1].start)
        {
            j--;
        }

        if (jobs[i - 1].weight + maxWeightArray[j] > maxWeightArray[i - 1])
        {
            maxWeightArray[i] = jobs[i - 1].weight + maxWeightArray[j];
        }
        else
        {
            maxWeightArray[i] = maxWeightArray[i - 1];
        }
    }
}

// Weighted Interval Scheduling
int main(int argc, char *argv[])
{
    int numInstances = 0;

    // scan the first line from stdin to get the number of instances
    fscanf(stdin, "%d\n", &numInstances);

    while (numInstances--)
    {
        int numJobs = 0;
        // scan the line to get number of pair of points
        fscanf(stdin, "%d\n", &numJobs);

        // an array of jobs
        struct Job *jobs = (struct Job *)malloc(numJobs * sizeof(struct Job));

        // read the jobs from stdin
        for (int i = 0; i < numJobs; i++)
        {
            fscanf(stdin, "%d %d %d\n", &jobs[i].start, &jobs[i].end, &jobs[i].weight);
        }

        // sort the jobs based on their end time using quick sort
        sortJobs(jobs, numJobs);

        // create a 1D array to store the total weight of jobs that can be scheduled to avoid recomputation
        long long *maxWeightArray = (long long *)malloc(numJobs * sizeof(long long) + 1);
        // initialize the first element of the array to 0
        maxWeightArray[0] = 0;

        // determine the maximum weight of jobs that can be scheduled using dynamic programming
        maxWeight(numJobs, jobs, maxWeightArray);

        // print the maximum weight of jobs that can be scheduled
        printf("%lld\n", maxWeightArray[numJobs]);

        // free the memory allocated for the jobs and maxWeightArray
        free(jobs);
        free(maxWeightArray);
    }

    return 0;
}

File: p7_rust/wis.rs
use std::io;

struct Job {
    start: i32,
    end: i32,
    weight: i32,
}

// using binary search to find the latest non-overlapping job
fn find_latest_non_overlapping_job(jobs: &[Job], index: usize) -> isize {
    let mut low: isize = 0;
    let mut high: isize = index as isize - 1;

    while low <= high {
        let mid = (low + high) / 2;
        if jobs[mid as usize].end <= jobs[index].start {
            if mid == index as isize - 1 || jobs[(mid + 1) as usize].end > jobs[index].start {
                return mid;
            }
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    -1
}

fn main() {
    let stdin = io::stdin();

    // read number of instances
    let mut input = String::new();
    stdin.read_line(&mut input).unwrap();
    let num_instances: i32 = input.trim().parse().unwrap();

    for _ in 0..num_instances {
        // read number of jobs
        let mut input = String::new();
        stdin.read_line(&mut input).unwrap();
        let num_jobs: i32 = input.trim().parse().unwrap();

        let mut jobs: Vec<Job> = Vec::with_capacity(num_jobs);
        for _ in 0..num_jobs {
            let mut job_input = String::new();
            stdin.read_line(&mut job_input).unwrap();
            let job_data: Vec<i32> = job_input
                .split_whitespace()
                .map(|x| x.parse().unwrap())
                .collect();
            jobs.push(Job {
                start: job_data[0],
                end: job_data[1],
                weight: job_data[2],
            });
        }

        // sort jobs by end time
        jobs.sort_by(|a, b| a.end.cmp(&b.end));

        // create an array to store max weight to avoid recomputation
        let mut max_weight_array = vec![0u64; num_jobs + 1];

        for i in 1..=num_jobs {
            let incl_weight = jobs[i - 1].weight as u64;
            let l = find_latest_non_overlapping_job(&jobs, i - 1);
            let incl_weight = if l != -1 { incl_weight + max_weight_array[l as usize + 1] } else { incl_weight };

            max_weight_array[i] = std::cmp::max(incl_weight, max_weight_array[i - 1]);
        }

        println!("{}", max_weight_array[num_jobs]);
    }
}